### PR TITLE
UnstableSrcUri: check for unstable SRC_URI

### PR DESCRIPTION
**Pull request by arthurzam**
I've currently collected the patterns from the linked issues. I can add more as requested.

For ::gentoo it results in:

```
dev-haskell/memory
  UnstableSrcUri: version 0.15.0-r1: unstable SRC_URI: [ https://patch-diff.githubusercontent.com/raw/vincenthz/hs-memory/pull/81.patch ]

dev-python/jupyter
  UnstableSrcUri: version 1.0.0-r4: unstable SRC_URI: [ https://patch-diff.githubusercontent.com/raw/jupyter/jupyter/pull/198.patch ]

net-voip/telepathy-rakia
  UnstableSrcUri: version 0.8.0-r1: unstable SRC_URI: [ https://patch-diff.githubusercontent.com/raw/TelepathyIM/telepathy-rakia/pull/1.patch ]
```

For ::guru it results in:

```
sci-electronics/verible
  UnstableSrcUri: version 0.0.2275: unstable SRC_URI: [ https://files.pythonhosted.org/packages/6b/34/415834bfdafca3c5f451532e8a8d9ba89a21c9743a0c59fbd0205c7f9426/six-1.15.0.tar.gz ]
  UnstableSrcUri: version 0.0.3124: unstable SRC_URI: [ https://files.pythonhosted.org/packages/6b/34/415834bfdafca3c5f451532e8a8d9ba89a21c9743a0c59fbd0205c7f9426/six-1.15.0.tar.gz ]
```

Resolves: https://github.com/pkgcore/pkgcheck/issues/505
Resolves: https://github.com/pkgcore/pkgcheck/issues/510

### CODE REVIEW

File: src/pkgcheck/checks/metadata.py [hunk 1]
```diff
@@ -1363,13 +1382,18 @@ def __init__(self, *args, use_addon):
             r"https?://(github\.com/.*?/.*?/archive/.+\.zip|"
             r"gitlab\.com/.*?/.*?/-/archive/.+\.zip)"
         )
+        self.unstable_uris = re.compile(
+            r"^https?://files.pythonhosted.org/packages/(?:[0-9a-fA-F]{2}/){2}|"
```
Comment (mgorny) [inline src/pkgcheck/checks/metadata.py:1386]: Err but these definitely aren't "unstable" — after all, they have precise checksum!

File: src/pkgcheck/checks/metadata.py [hunk 2]
```diff
@@ -1363,13 +1382,18 @@ def __init__(self, *args, use_addon):
...
+            r"^https?://patch-diff.githubusercontent.com/raw/.*/pull/[0-9]+.(patch|diff)$"
```
Comment (mgorny) [inline src/pkgcheck/checks/metadata.py:1387]: I suppose it's no longer "unstable" once PR is merged, though.

File: src/pkgcheck/checks/metadata.py (hunk 1 above)
Comment (arthurzam) [reply src/pkgcheck/checks/metadata.py:1386]: Oh, Then I've misunderstood https://github.com/pkgcore/pkgcheck/issues/505
How should I call that url?

Comment (arthurzam) [reply src/pkgcheck/checks/metadata.py]: removed it :)

File: src/pkgcheck/checks/metadata.py (hunk 2 above)
Comment (arthurzam) [reply src/pkgcheck/checks/metadata.py:1386]: I think this is fine. There aren't a lot of them in ::gentoo, and maybe there is a way to change a diff for merged PR?

Comment (thesamesam) [reply src/pkgcheck/checks/metadata.py:1386]: I don't think we can rely on it. See also https://github.com/orgs/community/discussions/56522.

File: testdata/repos/standalone/SrcUriCheck/UnstableSrcUri/UnstableSrcUri-0.ebuild [hunk 3]
```diff
@@ -0,0 +1,7 @@
+DESCRIPTION="Ebuild with unstable SRC_URI"
+HOMEPAGE="https://github.com/pkgcore/pkgcheck"
+SRC_URI="
+	https://patch-diff.githubusercontent.com/raw/pkgcore/pkgcheck/pull/599.patch
```
Comment (mgorny) [inline testdata/repos/standalone/SrcUriCheck/UnstableSrcUri/UnstableSrcUri-0.ebuild:4]: Will this check catch also the pre-redirect `https://github.com/foo/bar/pull/1234.patch` form? Perhaps we should do that as well.

Comment (arthurzam) [reply testdata/repos/standalone/SrcUriCheck/UnstableSrcUri/UnstableSrcUri-0.ebuild:4]: It didn't catch in original version, updated and included in tests.

**Review by mgorny — APPROVED** [review]
LGTM

**Pull request merged**